extensions/amp-analytics/0.1/events.js

```diff
@@ -33,6 +34,142 @@ const NO_UNLISTEN = function() {};
+
+/**
+ * Events that can result in analytics data to be sent.
+ * @const {!Object<string, {
+ *     name: string,
+ *     allowedFor: !Array<string>,
+ *     klass: function(new:./events.EventTracker)
+ *   }>}
+ */
+const TRACKER_TYPE = Object.freeze({
+  'click': {
+    name: 'click',
+    allowedFor: ALLOWED_FOR_ALL_ROOT_TYPES + ['timer'],
+    // Escape the temporal dead zone by not referencing a class directly.
+    klass: function(root) { return new ClickEventTracker(root); },
+  },
+  'custom': {
+    name: 'custom',
+    allowedFor: ALLOWED_FOR_ALL_ROOT_TYPES + ['timer'],
+    klass: function(root) { return new CustomEventTracker(root); },
+  },
+  'render-start': {
+    name: 'render-start',
+    allowedFor: ALLOWED_FOR_ALL_ROOT_TYPES + ['timer', 'visible'],
+    klass: function(root) { return new SignalTracker(root); },
+  },
+  'ini-load': {
+    name: 'ini-load',
+    allowedFor: ALLOWED_FOR_ALL_ROOT_TYPES + ['timer', 'visible'],
+    klass: function(root) { return new IniLoadTracker(root); },
+  },
+  'timer': {
+    name: 'timer',
+    allowedFor: ALLOWED_FOR_ALL_ROOT_TYPES,
+    klass: function(root) { return new TimerEventTracker(root); },
+  },
+  'visible': {
+    name: 'visible',
+    allowedFor: ALLOWED_FOR_ALL_ROOT_TYPES + ['timer'],
+    klass: function(root) { return new VisibilityTracker(root); },
+  },
+  'hidden': {
+    name: 'visible', // Reuse tracker with visibility
+    allowedFor: ALLOWED_FOR_ALL_ROOT_TYPES + ['timer'],
+    klass: function(root) { return new VisibilityTracker(root); },
+  },
+  'video': {
+    name: 'video',
+    allowedFor: ALLOWED_FOR_ALL_ROOT_TYPES + ['timer'],
+    klass: function(root) { return new VideoEventTracker(root); },
+  },
+});
+
+/**
+ * @param {string} triggerType
+ * @return {boolean}
+ */
+function isVideoTriggerType(triggerType) {
+  return startsWith(triggerType, 'video-');
+}
+
+/**
+ * @param {string} triggerType
+ * @return {boolean}
+ */
+function isReservedTriggerType(triggerType) {
+  return !!TRACKER_TYPE[triggerType] ||
+      !!isEnumValue(AnalyticsEventType, triggerType);
+}
+
+/**
+ * @param {string} eventType
+ * @return {string}
+ */
+export function getTrackerKeyName(eventType) {
+  if (isVideoTriggerType(eventType)) {
+    return 'video';
+  }
+  if (!isReservedTriggerType(eventType)) {
+    return 'custom';
+  }
+  return TRACKER_TYPE.hasOwnProperty(eventType) ?
+      TRACKER_TYPE[eventType].name : eventType;
+}
+
+/**
+ * @param {string} rootType
+ * @return {!Object<string, function(new:./events.EventTracker)>}
+ */
+export function getTrackerTypesForRootType(rootType) {
+  return filterTrackers(function(trackerProfile) {
+    return trackerProfile.allowedFor.indexOf(rootType) != -1;
+  });
+}
+
+function getTrackerTypesForVisibilityTracker() {
+  return filterTrackers(function(trackerProfile) {
+    return trackerProfile.allowedFor.indexOf('visible') != -1;
+  });
+}
+
+function getTrackerTypesForTimerEventTracker() {
```

Comment:
can we merge these three functions to one. And taking an array as a input instead?

Reply:
Done.